File: crime-analytics-frontend/src/services/predictionService.js
import axiosInstance from '../api/axiosInstance';

export const predictionService = {
  async getAllPredictions() {
    const response = await axiosInstance.get('/predictions');
    return response.data;
  },

  async getPredictionById(id) {
    const response = await axiosInstance.get(`/predictions/${id}`);
    return response.data;
  },

  async createPrediction(data) {
    // data: { crimeId, crimeType, city, state, predictedRisk, confidenceScore, predictionDate, description }
    const response = await axiosInstance.post('/predictions', data);
    return response.data;
  },

  async getByCity(city) {
    const response = await axiosInstance.get(`/predictions/city/${encodeURIComponent(city)}`);
    return response.data;
  },

  async getByState(state) {
    const response = await axiosInstance.get(`/predictions/state/${encodeURIComponent(state)}`);
    return response.data;
  },

  async getByType(crimeType) {
    const response = await axiosInstance.get(`/predictions/type/${encodeURIComponent(crimeType)}`);
    return response.data;
  },

  async getByRisk(predictedRisk) {
    const response = await axiosInstance.get(`/predictions/risk/${encodeURIComponent(predictedRisk)}`);
    return response.data;
  },

  async updatePrediction(id, data) {
    const response = await axiosInstance.put(`/predictions/${id}`, data);
    return response.data;
  },

  async deletePrediction(id) {
    const response = await axiosInstance.delete(`/predictions/${id}`);
    return response.data;
  }
};

File: crime-analytics-backend/src/main/java/com/crimeanalytics/service/PredictionService.java
package com.crimeanalytics.service;

import com.crimeanalytics.dto.request.PredictionRequest;
import com.crimeanalytics.dto.request.response.PredictionResponse;

import java.util.List;

public interface PredictionService {

    PredictionResponse createPrediction(PredictionRequest request);

    List<PredictionResponse> getAllPredictions();

    PredictionResponse getPredictionById(Long id);

    List<PredictionResponse> getPredictionsByCity(String city);

    List<PredictionResponse> getPredictionsByState(String state);

    List<PredictionResponse> getPredictionsByCrimeType(String crimeType);

    List<PredictionResponse> getPredictionsByRisk(String predictedRisk);

    PredictionResponse updatePrediction(Long id, PredictionRequest request);

    void deletePrediction(Long id);
}